### ISSUE TITLE
Missed dependency in node_modules

### ISSUE THREAD

**Issue by AndyTyurin** [yarnpkg/yarn#5306]
**Do you want to request a *feature* or report a *bug*?**
Bug.

**What is the current behavior?**
I'm trying to fetch npm module with dependencies such as 'babel-loader' for further usage by webpack and promise to have 'babel-loader' inside my project's node_modules directory, but instead, it has been placed in "npm_module"/node_modules directory.

**If the current behavior is a bug, please provide the steps to reproduce.**
1) Fetch npm module https://github.com/alfa-laboratory/arui-presets .
2) Run `yarn install`.
3) Go to node_modules, there are no 'babel-loader' npm module.

**What is the expected behavior?**
'babel-loader' should be inside "projectDir"/node_modules, but not inside "projectDir"/node_modules/arui-presets/node_modules

**Please mention your node.js, yarn and operating system version.**
MacOS El Capitan 10.11.4,
Yarn 1.3.5,
Node.js 8.9.4

Thanks!

**Comment by GAumala** [comment]
Hello! Making assumptions about the structure of `node_modules` is dangerous. You cannot guarantee the exact path of any dependency.

**Comment by AndyTyurin** [comment]
GAumaia, thanks for your message. So, maybe I'll clarify my problem, I can't import 'babel-loader' from my webpack config, because it's not placed in node_modules of my project.

As an example you can see babel-loader loading below:
https://github.com/alfa-laboratory/arui-presets/blob/master/webpack.base.js
I'm importing arui-presets library in my project, but after yarn installs modules, it can't find 'babel-loader'.

It works great when I'm using yarn v0.27.5, also 'babel-loader' is listened on 'node_modules' of the project.

**Comment by GAumala** [comment]
Yarn is not deterministic about the location of packages installed to 'node_modules'. From [this blog post](https://yarnpkg.com/blog/2017/05/31/determinism/): 
> But there’s one key piece of information that the yarn lockfile doesn’t contain and that’s the hoisting and location of each dependency in the tree. For example, given a yarn.lock it’s impossible to determine what are the top level dependencies unless you have it’s accompanying package.json. Even knowing the top level packages, we still cannot infer what hoisting position packages should be in.

If you really need `babel-loader`, just add it to your project's dev dependencies along with `arui-presets`. The only inconvenience is that you'll have to manually update `babel-loader` to the correct version every time you want to update `arui-preset`.

**Comment by AndyTyurin** [comment]
I understood, thank you for blog post, it makes sense. Yarn is not deterministic, but npm is it?
I'm asking this because when I'm installing by using of `npm install` it works well.

**Comment by torifat** [comment]
That is not expected behavior. Please use latest stable or RC. I tried your repo on both and it's working as expected. 

<img width="485" alt="screen shot 2018-02-12 at 8 44 14 pm" src="https://user-images.githubusercontent.com/208544/36090760-b278551e-1035-11e8-833e-7519c870bcfd.png">
<img width="491" alt="screen shot 2018-02-12 at 8 44 49 pm" src="https://user-images.githubusercontent.com/208544/36090764-b43c3d52-1035-11e8-8cd2-af43459d9227.png">

**Comment by AndyTyurin** [comment]
> That is not expected behavior. Please use latest stable or RC. I tried your repo on both and it's working as expected.

I assume, except `arui-presets` there are another libraries have using `babel-loader`. Maybe to avoid collisions, library has been placed in `arui-presets/node_modules` directory. It's why you didn't stuck with same issue. But it's just a thought.

**Comment by torifat** [comment]
No, transitive dependencies will hoist, so that it can save you from having same things multiple times. But, it will not push your direct dependencies downward. If it does for some reason then it's a bug.

**Comment by AndyTyurin** [comment]
Yes, hoist in a case when version is the same? Anyway one version should be hoisted in my opinion.
But in my case I didn't have anything in project node_modules, only in `arui-presets`.
I tried to clear cache, re-install dependencies many times, nothing works, but works fine if I use npm.

Haven't got any idea why you didn't stuck with it. I'll try out to test on bare project, maybe will find something, beside I'm going to read again about how "nm" should works.

**Right now I have solved my problem by adding babel-loader to my package.json manually.**

Thank you.
